tensorflow_probability/python/distributions/mvn_full_covariance.py

```diff
@@ -118,71 +123,219 @@ def __init__(self,
+                                                      is_square=True)
+                                    
+    with tf.name_scope(name, values=[loc] + covariance_matrix.graph_parents) as name:
+      dtype = dtype_util.common_dtype([loc, covariance_matrix],
+                                      preferred_dtype=tf.float32)
+
+      self._loc = tf.convert_to_tensor(loc, name="loc", dtype=dtype)
+      self._covariance_matrix = covariance_matrix
+
+    super(MultivariateNormalFullCovariance, self).__init__(
+        dtype=dtype,
+        reparameterization_type=reparameterization.FULLY_REPARAMETERIZED,
+        parameters=parameters,
+        graph_parents=[self._loc] + self._covariance_matrix.graph_parents,
+        name=name,
+        validate_args=validate_args,
+        allow_nan_stats=allow_nan_stats)
     self._parameters = parameters
+
+  @property
+  def loc(self):
+    """The location parameter of the distribution.
+
+    `loc` applies an elementwise shift to the distribution.
+
+    ```none
+    X ~ MultivariateNormalFullCovariance(loc=0, covariance=1)   # Identity covariance, zero shift.
+    Y = covariance.cholesky() @ X + loc
+    ```
+
+    Returns:
+      The `loc` `Tensor`.
+    """
+    return self._loc
+
+  @property
+  def covariance_matrix(self):
+    """The covariance parameter of the distribution.
+
+    `covariance_matrix` applies an affine covariance to the distribution.
+
+    ```none
+    X ~ MultivariateNormalFullCovariance(loc=0, covariance=1)   # Identity covariance, zero shift.
+    Y = covariance.cholesky() @ X + loc
+    ```
+
+    Returns:
+      The `covariance_matrix` `LinearOperator`.
+    """
+    return self._covariance_matrix
+
+  def _batch_shape_tensor(self):
+    shape_list = [
+        self._covariance_matrix.batch_shape_tensor(),
+        tf.shape(self.loc)[:-1]
+    ]
+    return functools.reduce(tf.broadcast_dynamic_shape, shape_list)
+
+  def _batch_shape(self):
+    shape_list = [self._covariance_matrix.batch_shape, self.loc.shape[:-1]]
+    return functools.reduce(tf.broadcast_static_shape, shape_list)
+
+  def _event_shape_tensor(self):
+    return self._covariance_matrix.range_dimension_tensor()[tf.newaxis]
+
+  def _event_shape(self):
+    return self._covariance_matrix.range_dimension
+
+  def _sample_shape(self):
+    return tf.concat([self.batch_shape_tensor(), self.event_shape_tensor()], -1)
+
+  def _mean(self):
+    return _broadcast_to_shape(self.loc, self._sample_shape())
+
+  def _mode(self):
+    return _broadcast_to_shape(self.loc, self._sample_shape())
+
+  def _variance(self):
+    return self._covariance_matrix.diag_part()
+
+  def _covariance(self):
+    return self._covariance_matrix.to_dense()
+
+  def _stddev(self):
+    return tf.sqrt(self._variance())
+
+  def _sample_n(self, n, seed=None):
+    seed = seed_stream.SeedStream(seed, salt="multivariate normal")
+
+    loc = _broadcast_to_shape(self.loc, self._sample_shape())
+    mvn = mvn_tril.MultivariateNormalTriL(
```

Comment:
it seems like this won't be able to (efficiently) sample a low rank + diag because of the cholesky.